cordapp-cpk/src/main/kotlin/net/corda/plugins/cpk/OsgiExtension.kt

```diff
@@ -0,0 +1,268 @@
+import java.util.Properties
+import java.util.StringJoiner
+
+/**
+ * Registers these [OsgiExtension] properties as task inputs,
+ * because Gradle cannot "see" their `@Input` annotations yet.
+ */
+fun TaskInputs.nested(nestName: String, osgi: OsgiExtension) {
+    property("${nestName}.autoExport", osgi.autoExport)
+    property("${nestName}.exports", osgi.exports)
+    property("${nestName}.embeddedJars", osgi.embeddedJars)
+    property("${nestName}.imports", osgi.imports)
+    property("${nestName}.scanCordaClasses", osgi.scanCordaClasses)
+    property("${nestName}.symbolicName", osgi.symbolicName)
+}
+
+@Suppress("UnstableApiUsage", "MemberVisibilityCanBePrivate", "unused")
+open class OsgiExtension(objects: ObjectFactory, jar: Jar) {
+    private companion object {
+        private const val CORDAPP_CONFIG_PLUGIN_ID = "net.corda.cordapp.cordapp-configuration"
+        private const val CORDAPP_CONFIG_FILENAME = "cordapp-configuration.properties"
+
+        private val CORDA_CLASSES = "^Corda-.+-Classes\$".toRegex()
+
+        private val BASE_CORDA_CLASSES: Map<String, String> = unmodifiableMap(mapOf(
+            CORDA_CONTRACT_CLASSES to "IMPLEMENTS;net.corda.v5.ledger.contracts.Contract",
+            CORDA_WORKFLOW_CLASSES to "IMPLEMENTS;net.corda.core.flows.Flow",
+            CORDA_MAPPED_SCHEMA_CLASSES to "EXTENDS;net.corda.core.node.services.persistence.MappedSchema",
+            CORDA_SERVICE_CLASSES to "IMPLEMENTS;net.corda.v5.serialization.SerializeAsToken;HIERARCHY_INDIRECTLY_ANNOTATED;net.corda.core.node.services.CordaService"
+        ))
+
+        /**
+         * We need to import these packages so that the OSGi framework
+         * will create bundle wirings for them. This allows Hibernate
+         * to create lazy proxies for any JPA entities inside the CPK.
+         *
+         * We DO NOT want to bind the CPK to use specific versions of
+         * either Hibernate or Javassist here.
+         */
+        private val BASE_REQUIRED_PACKAGES: Set<String> = unmodifiableSet(setOf(
+            "org.hibernate.annotations",
+            "org.hibernate.proxy"
+        ))
+
+        @Throws(IOException::class)
+        private fun loadConfig(input: InputStream): Map<String, String> {
+            return Properties().let { props ->
+                input.buffered().use(props::load)
+                props.entries.associate { entry ->
+                    entry.key.toString() to entry.value.toString().trim()
+                }
+            }
+        }
+
+        fun dashConcat(first: String, second: String?): String {
+            return if (second.isNullOrEmpty()) {
+                first
+            } else {
+                "$first-$second"
+            }
+        }
+
+        fun parsePackages(value: String): Set<String> {
+            return value.split(",").mapTo(LinkedHashSet(), String::trim)
+        }
+
+        fun dynamic(value: String): String = "$value;resolution:=dynamic;version=!"
```

Comment:
What does `version=!`  do?

Reply:
It deletes the version attribute, should there be one. I didn't want a "dynamic import" to include a version.